File: Java/0201-0250/239. Sliding Window Maximum.java
// https://leetcode.com/problems/sliding-window-maximum/

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int size = nums.length;
        int[] result = new int[size-k+1];
        Deque<Integer> storage = new ArrayDeque<>();

        // Add K length elements to Dequeue
        for(int index=0; index<k; index++){
            while(!storage.isEmpty() && nums[index] >= nums[storage.peekLast()])
                storage.pollLast();
                storage.offerLast(index);
        }

        // Sliding Window
        int resultIndex = 0;
        for(int index=k; index<size; index++){
            result[resultIndex++] = nums[storage.peekFirst()];

            // Remove Elements out of the window range
            while(!storage.isEmpty() && storage.peekFirst() < index-k+1)
                storage.pollFirst();

            // Remove Smaller Elements than the current num form the End of Deque
            while(!storage.isEmpty() && nums[index] >= nums[storage.peekLast()])
                storage.pollLast();
            
            storage.offerLast(index);
        }

        result[resultIndex] = nums[storage.peekFirst()];
        return result;
    }
}

File: Python/0201-0250/239. Sliding Window Maximum.py
# https://leetcode.com/problems/sliding-window-maximum/

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        size, result, storage = len(nums), [], []

        # Add K length elements to Dequeue
        for index in range(0, k):
            while len(storage) > 0 and nums[index] >= nums[storage[len(storage)-1]]:
                storage.pop()
            storage.append(index)

        # Sliding Window
        for index in range(k, size):
            result.append(nums[storage[0]])

            # Remove Elements out of the window range
            while len(storage) > 0 and storage[0] < index-k+1:
                storage.pop(0)

            # Remove Smaller Elements than the current num form the End of Deque
            while len(storage) > 0 and nums[index] >= nums[storage[len(storage)-1]]:
                storage.pop()
            
            storage.append(index)

        result.append(nums[storage[0]])
        return result
